internal/controller/accessrequest_controller.go

```diff
@@ -275,6 +323,120 @@ func (r *AccessRequestReconciler) getRoleTemplate(ctx context.Context, ar *api.A
+// It adds the TTL (Time-To-Live) duration, defined in the controller configuration,
+// to the creation timestamp of the AccessRequest.
+//
+// Parameters:
+// - ar: The AccessRequest object containing the creation timestamp.
+// - config: The controller configuration providing the TTL duration.
+//
+// Returns:
+// - The calculated expiration time as a time.Time object.
+func getTTLTime(ar *api.AccessRequest, config config.ControllerConfigurer) time.Time {
+	return ar.GetCreationTimestamp().Time.Add(config.ControllerAccessRequestTTL())
+}
+
+// doWithRetry attempts to execute the provided function `fn` with retries on conflict
+// errors. It ensures that the AccessRequest object is re-fetched on subsequent attempts
+// after the first failure.
+//
+// Parameters:
+// - ctx: The context for managing request deadlines and cancellations.
+// - ar: The AccessRequest object to operate on.
+// - fn: A function that performs the desired operation on the AccessRequest object.
+//
+// Returns:
+// - An error if the operation fails after exhausting all retries.
+func (r *AccessRequestReconciler) doWithRetry(ctx context.Context, ar *api.AccessRequest, fn func(ctx context.Context, ar *api.AccessRequest) error) error {
+	firstAttempt := true
+	return retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		if !firstAttempt {
+			// Re-fetch the AccessRequest object to ensure the latest state is used.
+			err := r.Get(ctx, client.ObjectKeyFromObject(ar), ar)
+			if err != nil {
+				return fmt.Errorf("error getting the AccessRequest: %s", err)
+			}
+		}
+		firstAttempt = false
+		// Execute the provided function with the AccessRequest object.
+		return fn(ctx, ar)
+	})
+}
+
+// handleTimeout checks if the AccessRequest resource has exceeded its configured timeout duration.
+// If the timeout is not configured, it returns false and does not update the status.
+// Should only process timeouts for AccessRequests in InitiatedStatus or in RequestedStatus.
+// If the timeout is exceeded, it updates the status to indicate a timeout.
+//
+// Parameters:
+// - ctx: The context for managing request deadlines and cancellations.
+// - ar: The AccessRequest object to evaluate.
+//
+// Returns:
+// - A boolean indicating whether the AccessRequest has timed out.
+// - An error if there is an issue updating the status.
+func (r *AccessRequestReconciler) handleTimeout(ctx context.Context, ar *api.AccessRequest) (bool, error) {
+	// If the timeout is not configured, return false and do not update the status.
+	if !hasTimeoutConfig(r.Config) {
+		return false, nil
+	}
+	// skip if not in InitiatedStatus or in RequestedStatus.
+	if ar.Status.RequestState != api.InitiatedStatus && ar.Status.RequestState != api.RequestedStatus {
+		return false, nil
+	}
+	// Check if the AccessRequest has exceeded its configured timeout duration.
+	timedout := time.Now().After(ar.GetCreationTimestamp().Time.Add(r.Config.ControllerRequestTimeout()))
```

Comment:
Should use the latest of the transition to initiated or requested as a start time instead of the creation date for the same reasons above.

Reply:
fixed